### PR TITLE
Handle the json extension when ResolvedModule result is json file

### LINKED ISSUE #24932: Errors when using `resolveJsonModule` through tools (due to missing `JSON` in the list of supported extensions?)
**TypeScript Version:**  `2.9.1`, `3.0.0-dev.20180609` with `resolveJsonModule: true`

**Search Terms:** `JSON`, `resolveJsonModule`, `import JSON`, `getOptionsDiagnostics`

**Code**

This only happens when using a tool that uses the API and not through `tsc` CLI, _but_ I tracked it down through to code in [TypeScript](https://github.com/Microsoft/TypeScript) itself (see repo & explanations below).

**Expected behavior:**

Compiles, parses `settings.json` (whatever JSON file) properly.

**Actual behavior:**

Gets `ts-loader` to throw `TypeError: Cannot read property 'getOptionsDiagnostics' of undefined` due to missing `program` (undefined) in `instance`.

Digging deeper, I can get to this error emitted in `TypeScript`: `Debug Failure. File /path/to/repo/settings.json has unknown extension.`

See https://github.com/TypeStrong/ts-loader/issues/793 for the full discussion until now and mostly [this comment](https://github.com/TypeStrong/ts-loader/issues/793#issuecomment-397012110) for the most TypeScript-related bit and a more complete explanation.

The issue can be solved by manually adding `".json"` to the `supportedTypescriptExtensionsForExtractExtension` list (in https://github.com/Microsoft/TypeScript/blob/master/src/parser/utilities.ts#L7812): compilation goes through and we actually get an output file that contains the JSON data as it should.

NB: we've come to suspect this bug might originate in `TypeScript`, but even if it is in `ts-loader`, any pointers as to a cause and/or possible fix would be very welcome.

**Playground Link:**

Here is a very minimal GitHub repository that reproduces this behavior: https://github.com/mbenadda/ts-loader-issue-793-example-repo

**Related Issues:**

N/A

### PR DESCRIPTION
**Pull request by sheetalkamat**
Fixes #24932

### CODE REVIEW

**Comment by johnnyreilly** [comment]
Great work!

File: src/harness/fourslash.ts [hunk 1]
```diff
@@ -507,8 +507,17 @@ namespace FourSlash {
...
         private getAllDiagnostics(): ts.Diagnostic[] {
-            return ts.flatMap(this.languageServiceAdapterHost.getFilenames(), fileName =>
-                ts.isAnySupportedFileExtension(fileName) ? this.getDiagnostics(fileName) : []);
+            return ts.flatMap(this.languageServiceAdapterHost.getFilenames(), fileName => {
+                if (!ts.isAnySupportedFileExtension(fileName)) {
+                    return [];
+                }
+
+                const baseName = ts.getBaseFileName(fileName);
+                if (baseName === "package.json" || baseName === "tsconfig.json" || baseName === "jsconfig.json") {
```
Comment (timocov) [inline src/harness/fourslash.ts]: > baseName === "tsconfig.json" 

But if we have tsconfig with custom name will it work?

**Comment by johnnyreilly** [comment]
Do you know when this is likely to be released?

File: src/harness/fourslash.ts (hunk 1 above)
Comment (liz282907) [reply src/harness/fourslash.ts]: have the same problem

Comment (timocov) [reply src/harness/fourslash.ts]: @sheetalkamat can you please take a look?

Comment (johnnyreilly) [reply src/harness/fourslash.ts]: I think you probably want to raise a new issue for this; I doubt people are looking at this as the original issue was solved and shipped with 3.0.0 AFAIK

**Pull request merged**